### PR TITLE
Get AccessKey from SignalR Service

### PR DESCRIPTION
**Pull request by terencefan**

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs [hunk 1]
```diff
@@ -156,5 +159,33 @@ private static (string, EndpointType) ParseKeyWithPrefix(string key, string pref
...
+                for (int i = 0; i < 3; i++)
+                {
+                    try
+                    {
+                        await key.AuthorizeAsync(Endpoint, Port, provider.GetName(), options);
+                        break;
+                    }
+                    catch (Exception)
+                    {
+                        // Retry authorizing access key, up to 3 times.
```
Comment (KKhurin) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:177]: Add log?

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs [hunk 2]
```diff
@@ -1,21 +1,79 @@
...
+
+        internal async Task AuthorizeAsync(string endpoint, int? port, string serverId, string accessToken)
+        {
+            var client = HttpClientFactory.CreateClient();
+            client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
+
+            var query = HttpUtility.ParseQueryString(string.Empty);
+            query.Add("serverId", serverId);
+
+            var builder = new UriBuilder(endpoint + "/api/v1/auth/accessKey")
```
Comment (KKhurin) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs:57]: (endpoint, "/api/v1/auth/accessKey")

File: samples/ChatSample/ChatSample/Startup.cs [hunk 3]
```diff
@@ -19,6 +19,12 @@ public void ConfigureServices(IServiceCollection services)
                 {
                     option.GracefulShutdown.Mode = GracefulShutdownMode.WaitForClientsClose;
                     option.GracefulShutdown.Timeout = TimeSpan.FromSeconds(10);
+
+                    option.AuthOptions = new AzureActiveDirectoryOptions(
+                        "70f09175-ecf3-477e-ad90-bb5dec839250",
+                        "tWI.0t.CT8iA2~e9vaQRXlkIB1VpYsAyb.",
+                        "c8a86907-dd80-4e5d-994d-36e0694e4913"
+                    );
```
Comment (terencefan) [inline samples/ChatSample/ChatSample/Startup.cs:27]: Test codes will remove later.

File: samples/ChatSample/ChatSample/appsettings.json [hunk 4]
```diff
@@ -2,7 +2,7 @@
   "Azure": {
     "SignalR": {
       "Enabled": true,
-      "ConnectionString": ""
+      "ConnectionString": "Endpoint=http://localhost;Version=1.0;Port=8080"
```
Comment (terencefan) [inline samples/ChatSample/ChatSample/appsettings.json:5]: Test codes will remove later.

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs [hunk 5]
```diff
@@ -2,24 +2,34 @@
...
+    public class ServiceEndpoint : IDisposable
     {
-        public string ConnectionString { get; }
+        private const int RetryTimes = 3;
 
-        public EndpointType EndpointType { get; }
+        private static readonly TimeSpan RetryDelay = TimeSpan.FromSeconds(1);
 
-        public virtual string Name { get; internal set; }
+        private readonly TimerAwaitable _timer = new TimerAwaitable(TimeSpan.Zero, TimeSpan.FromMinutes(55));
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:16]: every `HubServiceEndpoint` will have a `_timer` too

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs [hunk 6]
```diff
@@ -88,21 +96,13 @@ public ServiceEndpoint(ServiceEndpoint endpoint)
...
-            var prefix = string.IsNullOrEmpty(Name) ? "" : $"[{Name}]";
-            return $"{prefix}({EndpointType}){Endpoint}";
-        }
-
-        public override int GetHashCode()
-        {
-            // We consider ServiceEndpoint with the same Endpoint (https://{signalr.endpoint}) as the unique identity
-            return (Endpoint, EndpointType, Name).GetHashCode();
+            _updateAccessKeyTimer.Stop();
+            if (_updateAccessKeyTimer is IDisposable d)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:102]: >_updateAccessKeyTimer [](start = 16, length = 21)

`((IDisposable)_updateAccessKeyTimer).Dispose()`

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs [hunk 7]
```diff
@@ -146,6 +158,33 @@ internal static (string, EndpointType) ParseKey(string key)
...
+            {
+                try
+                {
+                    await key.AuthenticateAsync(Endpoint, Port, provider.GetName());
+                    return;
+                }
+                catch (Exception e)
+                {
+                    await Task.Delay(RetryDelay);
+                    logger.LogError("Failed to get an accesskey, retrying...", e);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:182]: >LogError [](start = 27, length = 8)

logWarning? this is inside customer code

File: src/Microsoft.Azure.SignalR.Management/ServiceManager.cs [hunk 8]
```diff
@@ -40,7 +40,7 @@ internal ServiceManager(ServiceManagerOptions serviceManagerOptions, string prod
             {
                 ApplicationName = _serviceManagerOptions.ApplicationName,
                 Proxy = serviceManagerOptions.Proxy
-            }).Value);
+            }).Value, new LoggerFactory());
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Management/ServiceManager.cs:43]: >new LoggerFactory() [](start = 22, length = 19)

does Management API support MSI? I would prefer using `NullLoggerFactory.Instance` instead

File: samples/ChatSample/ChatSample/Startup.cs [hunk 9]
```diff
@@ -14,13 +13,12 @@ public class Startup
...
-                    option.GracefulShutdown.Mode = GracefulShutdownMode.WaitForClientsClose;
-                    option.GracefulShutdown.Timeout = TimeSpan.FromSeconds(10);
-                })
-                .AddMessagePackProtocol();
+            services.AddSignalR().AddAzureSignalR(option =>
+            {
+                var authOptions = new AadManagedIdentityOptions();
+                var endpoint = new ServiceEndpoint("http://localhost", authOptions, port: 8080);
+                option.Endpoints = new ServiceEndpoint[] { endpoint };
+            }).AddMessagePackProtocol();
```
Comment (terencefan) [inline samples/ChatSample/ChatSample/Startup.cs:21]: Revert

**Pull request closed without merging**